pkg/live/inventoryrg.go

```diff
@@ -249,78 +308,432 @@ func (icm *InventoryResourceGroup) GetObject() (*unstructured.Unstructured, erro
+		return newRG
 	}
+
+	// Extract the index of the current sharded ResourceGroup to be incremented on.
+	str := strings.TrimPrefix(name, fmt.Sprintf("%s-%d-", icm.Name(), len(icm.Name())))
+	currID, _ := strconv.Atoi(str)
+	newRG.SetName(createShardedName(name, currID+1))
+	return newRG
+}
+
+// getObjCount returns the number of resources stored within the given ResourceGroup object.
+func getObjCount(rg *unstructured.Unstructured) int {
+	items, exists, err := unstructured.NestedSlice(rg.Object, "spec", "resources")
+	if err != nil || !exists {
+		return 0
+	}
+
+	return len(items)
+}
+
+// ApplyWithPrune applies the inventory objects with pruning logic.
+func (icm *InventoryResourceGroup) ApplyWithPrune(dc dynamic.Interface, mapper meta.RESTMapper, _ inventory.StatusPolicy, allObjs object.ObjMetadataSet) error {
+	namespacedClient, err := icm.getNamespacedClient(dc, mapper)
 	if err != nil {
 		return err
 	}
 
-	// Update status.
-	if statusPolicy == inventory.StatusPolicyAll {
-		invInfo.SetResourceVersion(appliedObj.GetResourceVersion())
-		_, err = namespacedClient.UpdateStatus(context.TODO(), invInfo, metav1.UpdateOptions{})
+	// Get all cluster ResourceGroup objects using label selector.
+	resourceList, err := namespacedClient.List(context.TODO(), metav1.ListOptions{
+		LabelSelector: fmt.Sprintf(shardedRGLabelFmt, icm.Name()),
+	})
+	if err != nil {
+		return err
 	}
 
-	return err
+	managedResources := make(map[object.ObjMetadata]struct{}, len(allObjs))
+	for _, id := range allObjs {
+		managedResources[id] = struct{}{}
+	}
+
+	needCollapse := false
+	for _, rg := range resourceList.Items {
+		items, exists, err := unstructured.NestedSlice(rg.Object, "spec", "resources")
+		if err != nil {
+			return err
+		}
+		if !exists {
+			continue
+		}
+
+		newItems := []interface{}{}
+		for _, item := range items {
+			fields, ok := item.(map[string]interface{})
+			if !ok {
+				return fmt.Errorf("unable to cast nested fields to the required types for: %v", item)
+			}
+
+			id, err := idFromFieldInterface(fields)
+			if err != nil {
+				return err
+			}
+
+			if _, exists := managedResources[id]; exists {
+				newItems = append(newItems, item)
+			} else {
+				needCollapse = true
+			}
+		}
+		err = unstructured.SetNestedSlice(rg.UnstructuredContent(), newItems, "spec", "resources")
+		if err != nil {
+			return err
+		}
+		_, err = namespacedClient.Update(context.TODO(), &rg, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	if needCollapse {
+		// Reshuffle GKNNs to the start of ResourceGroup object series.
+		return icm.collapse(dc, mapper)
+	}
+
+	return nil
 }
 
-func (icm *InventoryResourceGroup) ApplyWithPrune(dc dynamic.Interface, mapper meta.RESTMapper, statusPolicy inventory.StatusPolicy, _ object.ObjMetadataSet) error {
-	invInfo, namespacedClient, err := icm.getNamespacedClient(dc, mapper)
+// collapse is a utility function to rebalance objects stored within sharded ResourceGroups.
```

Comment:
collapse and rebalance mean different things to me. Perhaps this is more accurate: "collapse the set of ResourceGroups into as few ResourceGroups as possible, given the MaxResourceCount setting."